### PR TITLE
[ONNX] Cast scale back to fp16 after _attention_scale.

### PR DESCRIPTION
**Pull request by AdamLouly**
### **Description**:
The problem is that the graph was cast to `fp32` at a certain point but never reverted to `fp16`, causing the rest of the graph to run on `fp32`. This change aims to fix that issue and improve performance.

### **Changes Made**:
- Modified the ONNX exporter code to ensure that the graph is correctly cast back to `fp16` after a necessary cast to `fp32`.

### **Why This Change is Necessary**:
This change is necessary to ensure that the exported ONNX graph remains in `fp16` where appropriate, leading to significant gains in performance and memory savings. Without this fix, the graph would run entirely in `fp32`, causing suboptimal performance.

### **Testing**:
- Performed extensive testing with various models and scenarios to validate the correctness of the changes.

### **Benchmarking Results**:

Experiments Ran on: 
8 GPUS - Tesla V100 - 32GB

**Before Fix: ort + 4 hidden layers + without fix**

- **Train Runtime**: 78.7088 seconds
- **Train Samples per Second**: 10.164
- **Train Steps per Second**: 1.271
- **Train Loss**: 5.624655108451844
- **Epoch**: 0.3

**After Fix: ort + 4 hidden layers + with fix**

- **Train Runtime**: 72.5636 seconds
- **Train Samples per Second**: 11.025
- **Train Steps per Second**: 1.378
- **Train Loss**: 5.6252727746963505
- **Epoch**: 0.3

We can see 7.79% perf gain after this fix.

- I only ran it on 4 hidden layers due to GPU constraints, the perf gain is going to be much higher on the full model.
- You could see the gain on other models that uses _attention_scale as well.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

### CODE REVIEW

**Review by thiagocrepaldi — COMMENTED** [review]
Thanks for the contribution. While we wait CI to finish, please sign the [Linux Foundation SLA](https://contributor.easycla.lfx.linuxfoundation.org/#/cla/project/749a0166-38fe-4587-8b67-718d1d5d7ecf/user/af29016e-621b-44ec-9e8c-d94a18431cdd?redirect=https:%2F%2Fgithub.com%2Fpytorch%2Fpytorch%2Fpull%2F112554) to have the PR merged when the time comes

Sign as **individual** instead of *Corporate*

File: torch/onnx/symbolic_opset14.py [hunk 1]
```diff
@@ -238,6 +238,12 @@ def _attention_scale(
     )
     const_one = g.op("Constant", value_t=torch.tensor([1.0], dtype=torch.float))
     scale = g.op("Div", const_one, g.op("Sqrt", embedding_size))
+    # Add a Cast to convert the scale back to fp16
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset14.py:241]: nit: 
```suggestion
    # Add a Cast to convert the scale back to original dtype.
```

**Review by BowenBao — APPROVED** [review]

**Review by thiagocrepaldi — APPROVED** [review]

**Comment by AdamLouly** [comment]
Thanks for the review!
